### PR TITLE
Added multistage azdo pipeline

### PR DESCRIPTION
**Pull request by Salazander**
Added multistate AzDo pipeline with approval gates. Uses 3 variable groups to differentiate environments

### CODE REVIEW

File: .azdo/pipelines/generate-package.yml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    displayName: Generate packages for DEV environment
+    steps:
+    - script: src/build-package.sh 'package-history' 'output' "$(VERSION-ID)"
+      displayName: 'Generate packages locally'
+
+    - script: |
+        curl -u "$USERNAME:$PASSWORD" -XPUT "$(ARTIFACTORY-URL)/$(DEPENDENCY-PACK-NAME);deb.distribution=focal;deb.component=main;deb.architecture=amd64;deb.architecture=64bit-arm" -T "output/$(DEPENDENCY-PACK-NAME)" &&
+        curl -u "$USERNAME:$PASSWORD" -XPUT "$(ARTIFACTORY-URL)/$(DEPENDENCY-PACK-PINNING-NAME);deb.distribution=focal;deb.component=main;deb.architecture=amd64;deb.architecture=64bit-arm" -T "output/$(DEPENDENCY-PACK-PINNING-NAME)"
+      displayName: 'Upload packages to Artifactory'
+      env: # this is required for secret library variables
```
Comment (syedhassaanahmed) [inline .azdo/pipelines/generate-package.yml:33]: Should the Artifactory secret library variables (Url, Username and Password) be highlighted in the readme?

File: .azdo/pipelines/generate-package.yml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  - job: generate_packages_for_dev
+    displayName: Generate packages for DEV environment
+    steps:
+    - script: src/build-package.sh 'package-history' 'output' "$(VERSION-ID)"
+      displayName: 'Generate packages locally'
+
+    - script: |
+        curl -u "$USERNAME:$PASSWORD" -XPUT "$(ARTIFACTORY-URL)/$(DEPENDENCY-PACK-NAME);deb.distribution=focal;deb.component=main;deb.architecture=amd64;deb.architecture=64bit-arm" -T "output/$(DEPENDENCY-PACK-NAME)" &&
+        curl -u "$USERNAME:$PASSWORD" -XPUT "$(ARTIFACTORY-URL)/$(DEPENDENCY-PACK-PINNING-NAME);deb.distribution=focal;deb.component=main;deb.architecture=amd64;deb.architecture=64bit-arm" -T "output/$(DEPENDENCY-PACK-PINNING-NAME)"
+      displayName: 'Upload packages to Artifactory'
```
Comment (syedhassaanahmed) [inline .azdo/pipelines/generate-package.yml:32]: The 'upload packages to Artifactory' looks like a reusable script across the Dev, Test, Prod stages and can be extracted as an AzDO template. However, we can handle it in a separate PR.

**Review by syedhassaanahmed — APPROVED** [review]

File: .azdo/pipelines/generate-package.yml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  jobs:
+  - job: generate_packages_for_dev
+    displayName: Generate packages for DEV environment
+    steps:
+    - script: src/build-package.sh 'package-history' 'output' "$(VERSION-ID)"
+      displayName: 'Generate packages locally'
+
+    - script: |
+        curl -u "$USERNAME:$PASSWORD" -XPUT "$(ARTIFACTORY-URL)/$(DEPENDENCY-PACK-NAME);deb.distribution=focal;deb.component=main;deb.architecture=amd64;deb.architecture=64bit-arm" -T "output/$(DEPENDENCY-PACK-NAME)" &&
```
Comment (MagdaPaj) [inline .azdo/pipelines/generate-package.yml:30]: `ARTIFACTORY-URL` will it be the same for all 3 environments? Shouldn't each environment have unique URL and maybe also username, password?

File: .azdo/pipelines/generate-package.yml [hunk 4]
```diff
@@ -0,0 +1,101 @@
+trigger:
+  tags:
+    include:
+    - '*'
```
Comment (MagdaPaj) [inline .azdo/pipelines/generate-package.yml:4]: So this pipeline expect that a tag is pushed. Wouldn't be nicer to also generate and push a tag in the pipeline? And the pipeline would just expect a version parameter from a user, and not that a user pushes the tag.

File: .azdo/pipelines/generate-package.yml (hunk 3 above)
Comment (Salazander) [reply .azdo/pipelines/generate-package.yml:30]: every stage uses a different variable group (group: {DEV,TEST,PROD}). The variables are expected to be named to the same for the overriding logic to work

File: .azdo/pipelines/generate-package.yml (hunk 2 above)
Comment (Salazander) [reply .azdo/pipelines/generate-package.yml:32]: I will do this in a later PR

File: .azdo/pipelines/generate-package.yml (hunk 4 above)
Comment (Salazander) [reply .azdo/pipelines/generate-package.yml]: made tagging part of the pipeline

**Review by MagdaPaj — APPROVED** [review]

File: .azdo/pipelines/generate-package.yml (hunk 1 above)
Comment (Salazander) [reply .azdo/pipelines/generate-package.yml]: done, added details to README.md

**Pull request merged**